vms/platformvm/network/warp.go

```diff
@@ -0,0 +1,336 @@
+	subnetID, err := ids.ToID(justification)
+	if err != nil {
+		return &common.AppError{
+			Code:    ErrFailedToParseJustification,
+			Message: "failed to parse justification: " + err.Error(),
+		}
+	}
+
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	conversion, err := s.state.GetSubnetConversion(subnetID)
+	if err == database.ErrNotFound {
+		return &common.AppError{
+			Code:    ErrConversionDoesNotExist,
+			Message: fmt.Sprintf("subnet %q has not been converted", subnetID),
+		}
+	}
+	if err != nil {
+		return &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: "failed to get subnet conversionID: " + err.Error(),
+		}
+	}
+
+	if msg.ID != conversion.ConversionID {
+		return &common.AppError{
+			Code:    ErrMismatchedConversionID,
+			Message: fmt.Sprintf("provided conversionID %q != expected conversionID %q", msg.ID, conversion.ConversionID),
+		}
+	}
+
+	return nil
+}
+
+func (s signatureRequestVerifier) verifySubnetValidatorRegistration(
+	msg *message.SubnetValidatorRegistration,
+	justificationBytes []byte,
+) *common.AppError {
+	if msg.Registered {
+		return s.verifySubnetValidatorRegistered(msg.ValidationID)
+	}
+
+	var justification platformvm.SubnetValidatorRegistrationJustification
+	if err := proto.Unmarshal(justificationBytes, &justification); err != nil {
+		return &common.AppError{
+			Code:    ErrFailedToParseJustification,
+			Message: "failed to parse justification: " + err.Error(),
+		}
+	}
+
+	switch preimage := justification.GetPreimage().(type) {
+	case *platformvm.SubnetValidatorRegistrationJustification_ConvertSubnetTxData:
+		return s.verifySubnetValidatorNotCurrentlyRegistered(msg.ValidationID, preimage.ConvertSubnetTxData)
+	case *platformvm.SubnetValidatorRegistrationJustification_RegisterSubnetValidatorMessage:
+		return s.verifySubnetValidatorCanNotValidate(msg.ValidationID, preimage.RegisterSubnetValidatorMessage)
+	default:
+		return &common.AppError{
+			Code:    ErrInvalidJustificationType,
+			Message: fmt.Sprintf("invalid justification type: %T", justification.Preimage),
+		}
+	}
+}
+
+// verifySubnetValidatorCanNotValidate verifies that the validationID is
+// currently a validator.
+func (s signatureRequestVerifier) verifySubnetValidatorRegistered(
+	validationID ids.ID,
+) *common.AppError {
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	// Verify that the validator exists
+	_, err := s.state.GetSubnetOnlyValidator(validationID)
+	if err == database.ErrNotFound {
+		return &common.AppError{
+			Code:    ErrValidationDoesNotExist,
+			Message: fmt.Sprintf("validation %q does not exist", validationID),
+		}
+	}
+	if err != nil {
+		return &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: "failed to get subnet only validator: " + err.Error(),
+		}
+	}
+	return nil
+}
+
+// verifySubnetValidatorCanNotValidate verifies that the validationID is not
+// currently a validator.
+func (s signatureRequestVerifier) verifySubnetValidatorNotCurrentlyRegistered(
```

Comment:
Maybe we should rename this `verifyInitialSubnetValidatorRemoved`? It can't be used to verify any Subnet validator is not currently registered, only those validators added in the `ConverSubnetToL1Tx`.

Reply:
I think that `verifySubnetValidatorNotCurrentlyRegistered` is a bit more descriptive of the actual verification that is being performed than `verifyInitialSubnetValidatorRemoved`.

For example, if the `ConverSubnetToL1Tx` only registered 1 validator, then I would expect `verifyInitialSubnetValidatorRemoved` only to allow signing `subnetID.Append(0)`. But this function actually allows signing `subnetID.Append(1)` as well.

@michaelkaplan13 is this behavior acceptable? When I was writing this I had felt like it was similar to the case where a `SubnetValidatorRegistration` with `false` can be sent even though the subnet never requested it.

Reply:
Fair enough. Definitely agree with the behavior, don't love the name but see how my suggestion would be less accurate.